enhancements/storage/csi-vsphere-operator.md

```diff
@@ -0,0 +1,162 @@
+configured in OCP - such clusters will come up with no default storageClass installed and will require manual creation of storageClasses.
+
+### Deployment strategy
+
+vSphere CSI driver operator will be deployed by [cluster-storage-operator](https://github.com/openshift/cluster-storage-operator/)
+
+#### Operator deployment via cluster-storage-operator
+
+The cluster-storage-operator will deploy all the resources necessary for creating the vSphere CSI driver operator.
+
+1. vSphere CSI driver operator will be deployed in namespace `openshift-cluster-csi-drivers`.
+2. A service account will be created for running the operator.
+3. The operator will get RBAC rules necessary for running the operator and its operand (i.e the actual vSphere CSI driver).
+4. A deployment will be created for starting the operator and will be managed by cluster-storage-operator.
+5. A instance of `ClusterCSIDriver` will be created to faciliate managment of driver operator.  `ClusterCSIDriver` is already defined in - https://github.com/openshift/api/blob/master/operator/v1/types_csi_cluster_driver.go but needs to be expanded to include vSphere CSI driver.
+6. cluster-storage-operator will create required cloud-credentials for talking with vCenter API.
+
+#### Driver deployment via vsphere-csi-driver-operator.
+
+The operator itself will be responsible for running the driver and all the required sidecars (attacher, provisioner etc).
+
+1. The operator will create namespace `openshift-cluster-csi-drivers` to deploy the driver and sidecars.
+2. A service account will be created for running the driver.
+3. The operator will create RBAC rules necessary for running the driver and sidecars.
+4. A deployment will be created and managed by operator to handle control-plane sidecars and controller-plane driver deployment with controller-side of CSI services.
+5. A DaemonSet will be created and managed by operator to run node side of driver operations.
+6. A `CSIDriver` object will be created to expose driver features to control-plane.
+7. The driver operator will use and expose cloud-credentials created by cluster-storage-operator.
+
+Most of the steps outlined above is common to all CSI driver operator and vSphere CSI driver is not unique among those aspects.
+
+#### Additional consideration for vSphere
+
+There are certain aspects of driver which require special handling:
+
+1. When vSphere CSI operator starts, using credentials provided by cluster-storage-operator, it will first verifiy vCenter version and HW version of VMs. If vCenter version is not 6.7u3 or greater and HW version is not 15 or greater on all VMs - it will set `vSphereOperatorDisabled: true` and stop further processing. If additional VMs are added later into the cluster and they do not have HW version 15 or greater, Operator will mark itself as `degraded` and nodes which don't have right HW version will have annotation `vsphere.driver.status.csi.openshift.io: degraded` added to them.
```

Comment:
Using a HW with ver < 15  should initiate an alert, but the cluster should not get degraded, at least in 4.7.

4.7 *recommends* using HW ver. 15, but it works with older version. At some point we will *require* ver. 15, but exact OCP version is not set yet. Users need to update the VMs manually and fix any issues with it. Good documentation is required.

Reply:
We can check vCenter version and all ESXi host versions for the specific vSphere cluster that is configured in the install-config.yaml. Not 100% sure what the appropriate outcome of that information gathering should be though.

Reply:
I think we want uniform behaviour between operator noticing a VM with <=15 later vs it notices a VM afterwards. In both cases, we ideally want same thing to happen. 

I am thinking of snowflaking this degraded status in CSO, so as we don't degrade the cluster.

Scenario #1. vSphere operator starts and notices one or more VMs are not on HW >=15. It marks itself as degraded but starts controller and daemonset pods anyways. nodes which don't have matching HW version are labelled.

Scenario#2. vSphere notices that there are VMs which are not on HW >=15 during normal course of operation. It marks itself as degraded but  nodes which don't have matching HW version will have driver pods running. 

@huffmanca this also kinda addresses point you raised above. I don't think it is a good idea to treat these two cases as different ones. I do understand, this will result in pods silently failing to come up or if controller side of driver got scheduled on a node with HW <=15 then attach/detach and volume provisioning won't work. But if we base our scheduling decisions around HW on nodes, it will get very complicated..